src/uu/cksum/src/cksum.rs

```diff
@@ -391,47 +394,170 @@ fn calculate_length(algo_name: &str, length: usize) -> UResult<Option<usize>> {
  * and "easier" to understand
  */
-fn handle_tag_text_binary_flags(matches: &clap::ArgMatches, check: bool) -> UResult<(bool, bool)> {
+fn handle_tag_text_binary_flags(matches: &clap::ArgMatches) -> UResult<(bool, bool)> {
     let untagged: bool = matches.get_flag(options::UNTAGGED);
     let tag: bool = matches.get_flag(options::TAG);
     let tag: bool = tag || !untagged;
 
-    let text_flag: bool = matches.get_flag(options::TEXT);
     let binary_flag: bool = matches.get_flag(options::BINARY);
 
     let args: Vec<String> = std::env::args().collect();
     let had_reset = had_reset(&args);
 
     let asterisk: bool = prompt_asterisk(tag, binary_flag, had_reset);
 
-    if (binary_flag || text_flag) && check {
+    Ok((tag, asterisk))
+}
+
+/***
+ * Do the checksum validation (can be strict or not)
+*/
+fn perform_checksum_validation<'a, I>(files: I, strict: bool) -> UResult<()>
+where
+    I: Iterator<Item = &'a OsStr>,
+{
+    let re = Regex::new(r"(?P<algo>\w+)(-(?P<bits>\d+))? \((?P<filename>.*)\) = (?P<checksum>.*)")
+        .unwrap();
+    let mut properly_formatted = false;
+    let mut bad_format = 0;
+    let mut failed_cksum = 0;
+    let mut failed_open_file = 0;
+    for filename_input in files {
+        let input_is_stdin = filename_input == OsStr::new("-");
+        let file: Box<dyn Read> = if input_is_stdin {
+            Box::new(stdin()) // Use stdin if "-" is specified
+        } else {
+            match File::open(filename_input) {
+                Ok(f) => Box::new(f),
+                Err(err) => {
+                    show!(err.map_err_context(|| format!(
+                        "Failed to open file: {}",
+                        filename_input.to_string_lossy()
+                    )));
+                    return Err(io::Error::new(io::ErrorKind::Other, "Failed to open file").into());
+                }
+            }
+        };
+        let reader = BufReader::new(file);
+
+        for line in reader.lines() {
+            let line = line?;
+
+            if let Some(caps) = re.captures(&line) {
+                properly_formatted = true;
+                let algo_name = caps.name("algo").unwrap().as_str().to_lowercase();
+                let filename_to_check = caps.name("filename").unwrap().as_str();
+                let expected_checksum = caps.name("checksum").unwrap().as_str();
+
+                let length = caps
+                    .name("bits")
+                    .map(|m| m.as_str().parse::<usize>().unwrap() / 8);
+                let (_, mut algo, bits) = detect_algo(&algo_name, length);
+
+                let file_to_check: Box<dyn Read> = if filename_to_check == "-" {
+                    Box::new(stdin()) // Use stdin if "-" is specified in the checksum file
+                } else {
+                    match File::open(filename_to_check) {
+                        Ok(f) => Box::new(f),
+                        Err(err) => {
```

Comment:
In this case we should also write `FAILED` to stdout, in case a script only monitors that fd (and not stderr). It's what GNU does, too:

```console
$ echo 'MD5 (nonexistent) = e5773576fc75ff0f8eba14f61587ae28' > foo.sums 
$ ../gnu/src/cksum -c foo.sums 2> /dev/null
nonexistent: FAILED open or read
[$? = 1]
$ cargo run -q cksum -c foo.sums 2> /dev/null
[$? = 1]
```